index.js

```diff
@@ -1,105 +1,244 @@
-// Assert if `test` passes for `node`.
-// When a `parent` node is known the `index` of node should also be given.
-// eslint-disable-next-line max-params
-export function is(node, test, index, parent, context) {
-  var check = convert(test)
-
-  if (
-    index !== undefined &&
-    index !== null &&
-    (typeof index !== 'number' ||
-      index < 0 ||
-      index === Number.POSITIVE_INFINITY)
-  ) {
-    throw new Error('Expected positive finite index')
-  }
-
-  if (
-    parent !== undefined &&
-    parent !== null &&
-    (!is(parent) || !parent.children)
-  ) {
-    throw new Error('Expected parent node')
-  }
-
-  if (
-    (parent === undefined || parent === null) !==
-    (index === undefined || index === null)
-  ) {
-    throw new Error('Expected both parent and index')
-  }
-
-  return node && node.type && typeof node.type === 'string'
-    ? Boolean(check.call(context, node, index, parent))
-    : false
-}
-
-export function convert(test) {
-  if (test === undefined || test === null) {
-    return ok
-  }
+/**
+ * @typedef {import('unist').Node} Node
+ * @typedef {import('unist').Parent} Parent
+ *
+ * @typedef {string} Type
+ * @typedef {Object<string, unknown>} Props
+ */
+
+/**
+ * Check if a node passes a test
+ *
+ * @callback TestFunctionAnything
+ * @param {Node} node
+ * @param {number} [index]
+ * @param {Parent} [parent]
+ * @returns {boolean|void}
+ */
+
+/**
+ * Check if a node passes a certain node test
+ *
+ * @template {Node} X
+ * @callback TestFunctionPredicate
+ * @param {Node} node
+ * @param {number} [index]
+ * @param {Parent} [parent]
+ * @returns {node is X}
+ */
+
+/**
+ * @callback AssertAnything
+ * @param {unknown} [node]
+ * @param {number} [index]
+ * @param {Parent} [parent]
+ * @returns {boolean}
+ */
+
+/**
+ * Check if a node passes a certain node test
+ *
+ * @template {Node} Y
+ * @callback AssertPredicate
+ * @param {unknown} [node]
+ * @param {number} [index]
+ * @param {Parent} [parent]
+ * @returns {node is Y}
+ */
+
+/**
+ * Check if a node passes a test.
+ * When a `parent` node is known the `index` of node should also be given.
+ *
+ * @type {(
+ *   (<T extends Node>(node: unknown, test?: null|undefined, index?: number, parent?: Parent, context?: unknown) => node is Node) &
```

Comment:
This one seems a bit weird to me (but it was already supported): the type parameter isn’t used: `node` is instead checked for being any unist node. Should this remain supported?

Reply:
this is close, I believe the assertion is/was

```suggestion
 *   (<T extends Node>(node: unknown, test?: null|undefined, index?: number, parent?: Parent, context?: unknown) => node is T) &
```

where the generic `T` is used in the predicate, and asserts that `T` must extend `Node`, but can be more specific than `Node`

Reply:
Isn’t it unexpected that `<MyNode>is(someNode)` does a TS check but not the runtime check?

Reply:
Ah I see, this is the null predicate.
In that case the generic can be dropped.

Reply:
done!